econml/utilities.py

```diff
@@ -697,3 +1021,435 @@ def predict(self, Xt):
         t = Xt[:, -self.n_T:]
         predictions = [self.model_list[np.nonzero(t[i])[0][0]].predict(X[[i]]) for i in range(len(X))]
         return np.concatenate(predictions)
+
+
+def _safe_norm_ppf(q, loc=0, scale=1):
+    prelim = loc.copy()
+    if hasattr(prelim, "__len__"):
+        if np.any(scale > 0):
+            prelim[scale > 0] = scipy.stats.norm.ppf(q, loc=loc[scale > 0], scale=scale[scale > 0])
+    elif scale > 0:
+        prelim = scipy.stats.norm.ppf(q, loc=loc, scale=scale)
+    return prelim
+
+
+class StatsModelsLinearRegression:
+    """
+    Class which mimics weighted linear regression from the statsmodels package.
+
+    However, unlike statsmodels WLS, this class also supports sample variances in addition to sample weights,
+    which enables more accurate inference when working with summarized data.
+
+    Parameters
+    ----------
+    fit_intercept : bool (optional, default=True)
+        Whether to fit an intercept in this model
+    fit_args : dict (optional, default=`{}`)
+        The statsmodels-style fit arguments; keys can include 'cov_type', 'cov_kwds', and 'use_t'.
+    """
+
+    def __init__(self, fit_intercept=True, fit_args={}):
+        self.fit_args = fit_args
+        self._fit_intercept = fit_intercept
+        return
+
+    def _check_input(self, X, y, sample_weight, sample_var):
+        """Check dimensions and other assertions."""
+        if sample_weight is None:
+            sample_weight = np.ones(X.shape[0])
+        elif np.any(np.not_equal(np.mod(sample_weight, 1), 0)):
+            raise AttributeError("Sample weights must all be integers for inference to be valid!")
+
+        if sample_var is None:
+            if np.any(np.not_equal(sample_weight, 1)):
+                warnings.warn(
+                    """No variance information was given for samples with sample_weight not equal to 1,
+                       that represent summaries of multiple original samples. Inference will be invalid!""")
+            sample_var = np.zeros(y.shape)
+
+        if sample_var.ndim < 2:
+            if np.any(np.equal(sample_weight, 1) & np.not_equal(sample_var, 0)):
+                warnings.warn(
+                    """Variance was set to non-zero for an observation with sample_weight=1!
```

Comment:
These inference related warnings are raised even if the user had set `inference=None` when calling the cate estimator. This means that a user might have just wanted to fit a weighted point prediction, for which sample variances are not required and would have still got a warning. Seems ok for now, but maybe sth to thing if we want to control better when to display.